File: sequencial.py
from PIL import Image
import numpy as np
import time  # Importa o módulo para medir o tempo

def sequencial(path, intervalos):
    relatorio = {}

    imagem = Image.open(path).convert("L")  # carregar a imagem e converter pra grayscale
    matriz = np.array(imagem)  # transformar a imagem em um array NumPy

    tamanho_intervalo = 256 / intervalos  # tamanho de cada intervalo
    histograma = np.zeros(intervalos, dtype=int)

    linhas, colunas = matriz.shape

    inicio = time.time()  # tempo inicial

    for linha in range(linhas):
        for coluna in range(colunas):
            luminescencia = matriz[linha][coluna]
            intervalo_escolhido = int(luminescencia / tamanho_intervalo)
            if intervalo_escolhido >= intervalos:
                intervalo_escolhido = intervalos - 1  # caso 255 caia exatamente no último intervalo
            histograma[intervalo_escolhido] += 1

    fim = time.time()  # tempo final

    tempo_execucao_ms = (fim - inicio) * 1000  # calcula tempo em ms

    relatorio["histograma"] = histograma
    relatorio["tempo_ms"] = tempo_execucao_ms

    return relatorio

File: sequencial.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <chrono>
#include <vector>
#include <cuda_runtime.h>

using namespace std;
using namespace cv;


void sequencial(string imagePath, int intervalos) {
    Mat imagem = imread(imagePath, IMREAD_GRAYSCALE);

    int linhas = imagem.rows;
    int colunas = imagem.cols;
    int tamanho_intervalo = 256 / intervalos;

    vector<int> histograma(intervalos, 0);

    //tempo inicial
    auto inicio = chrono::high_resolution_clock::now();

    for (int linha = 0; linha < linhas; linha++) {
        for (int coluna = 0; coluna < colunas; coluna++) {
            int luminescencia = imagem.at<uchar>(linha, coluna);//metodo pra acessar coordenada especifica do OpenCV
            int intervalo_escolhido = luminescencia / tamanho_intervalo;
            if (intervalo_escolhido >= intervalos)
                intervalo_escolhido = intervalos - 1;
            histograma[intervalo_escolhido]++;
        }
    }

    auto fim = chrono::high_resolution_clock::now();//tempo final
    chrono::duration<double, milli> delta = fim - inicio;//delta tempo

    cout << "Histograma (Sequencial)" << endl;
    for (int i = 0; i < intervalos; i++) {
        cout << i << ": " << histograma[i] << endl;
    }

    cout << "Tempo: " << delta.count() << " ms\n\n\n" << endl;
}


int main() {
    sequencial("imagens/32k.png", 16);
    return 0;
}
